clang/lib/Sema/SemaDecl.cpp

```diff
@@ -15609,6 +15661,10 @@ Decl *Sema::ActOnStartOfFunctionDef(Scope *FnBodyScope, Decl *D,
       getCurLexicalContext()->getDeclKind() != Decl::ObjCImplementation)
     Diag(FD->getLocation(), diag::warn_function_def_in_objc_container);
 
+  if (Context.hasAnyFunctionEffects())
+    if (const auto FX = FD->getFunctionEffects(); !FX.empty())
+      maybeAddDeclWithEffects(FD, FX);
```

Comment:
This check is duplicate three times. Can we move it into `maybeAddDeclWithEffects`?

Reply:
Not sure if you saw the performance PR where `hasAnyFunctionEffects()` was introduced but it's left me a little paranoid about even just the function call overhead. Possibly a good compromise would be to make `maybeAddDeclWithEffects` inline.

Reply:
> even just the function call overhead

Hmm, it shouldn’t be *that* bad, but also, I haven’t benchmarked it. This is called once per function declaration, so it shouldn’t be too bad, imo. That said, making it inline sounds fine too because it’s fairly short.

Reply:
Factored this 3-line sequence into an inline method of `Sema` (which has to be templated to accommodate `FunctionDecl` vs. `BlockDecl`).